static_frame/core/index_hierarchy_set_utils.py

```diff
@@ -0,0 +1,480 @@
+import typing as tp
+from functools import partial
+
+import numpy as np
+from arraykit import mloc
+
+from static_frame.core.container_util import index_many_to_one
+from static_frame.core.exception import ErrorInitIndex
+from static_frame.core.index import Index
+from static_frame.core.index_hierarchy import IndexHierarchy
+from static_frame.core.loc_map import HierarchicalLocMap
+from static_frame.core.util import DtypeSpecifier
+from static_frame.core.util import IndexConstructor
+from static_frame.core.util import ManyToOneType
+from static_frame.core.util import intersect1d
+from static_frame.core.util import setdiff1d
+from static_frame.core.util import ufunc_unique1d
+from static_frame.core.util import ufunc_unique1d_indexer
+
+
+class ValidationResult(tp.NamedTuple):
+    indices: tp.List[IndexHierarchy]
+    depth: int
+    any_dropped: bool
+    any_shallow_copies: bool
+    name: tp.Hashable
+    index_constructors: tp.List[IndexConstructor]
+
+
+def _get_shallow_copy_key(ih: IndexHierarchy) -> tp.Hashable:
+    # calling id(...) on numpy arrays is not always reliable, so instead,
+    # we will use the underlying array properties to determine the ID for this
+    # index hierarchy, with the desire to encounter duplicate keys for shallow copies
+
+    result: tp.List[tp.Union[int, tp.Tuple[int, ...]]] = []
+
+    for block in ih._blocks._blocks:
+        # You need all three of these to uniquely identify a numpy array
+        result.append(mloc(block))
+        result.append(block.shape)
+        result.append(block.strides)
+
+    return tuple(result)
+
+
+def _validate_and_process_indices(
+        indices: tp.Tuple[IndexHierarchy],
+        ) -> ValidationResult:
+    '''
+    Common sanitization for IndexHierarchy operations.
+    1. Remove empty or cloned indices
+    2. Ensure all indices have same depth
+    3. Use the first index's name, if all indices have same name
+    4. Use the first index's index_constructors, if all indices have same index_constructors
+
+    This will also invoke recache on all indices due to the `.size` call
+    '''
+    any_dropped = False
+    any_shallow_copies = False
+
+    name: tp.Hashable = indices[0].name
+    index_constructors = list(indices[0]._index_constructors)
+
+    shallow_copy_keys: tp.Set[tp.Hashable] = set()
+    unique_non_empty_indices: tp.List[IndexHierarchy] = []
+
+    depth: tp.Optional[int] = None
+    for idx in indices:
+        if name is not None and idx.name != name:
+            name = None
+
+        for ctor, other_ctor in zip(idx._index_constructors, index_constructors):
+            if other_ctor != ctor:
+                index_constructors = []
+                break
```

Comment:
@flexatone I figured this was unnecessary, since if `index_constructors` is empty, then the `zip` iterable will also be empty.

I was trying to avoid the extra branch. What do you think?

Reply:
Sure, I am ambivalent. The extra branch avoids the `zip` call and empty iteration, but I agree it is of questionable benefit. It does, however, make the logic a little more clear.